pkg/querier/queryrange/series_cache.go

```diff
@@ -0,0 +1,100 @@
+package queryrange
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sort"
+	strings "strings"
+	"time"
+
+	"github.com/go-kit/log"
+
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/storage/chunk/cache/resultscache"
+)
+
+type cacheKeySeries struct {
+	Limits
+	transformer UserIDTransformer
+}
+
+// GenerateCacheKey generates a cache key based on the userID, matchers, split duration and the interval of the request.
+func (i cacheKeySeries) GenerateCacheKey(ctx context.Context, userID string, r resultscache.Request) string {
+	sr := r.(*LokiSeriesRequest)
```

Comment:
Similar to what @ashwanthgoli mentioned above.

1. If this panics, it's almost always means we chained wrong middleware for series request.
2. It would break `GenerateCacheKey` api interface, only thing we can do with catched error is to panic, which we already do now.